### ISSUE TITLE
Issues with ld and num when building with Dune 2 within nix-shell.

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#11333]
I tried to update the Dune version in `default.nix` to make it possible again to compile Coq with Dune from within a `nix-shell`:

```diff
diff --git a/default.nix b/default.nix
index ee4a6046ea..8221c38502 100644
--- a/default.nix
+++ b/default.nix
@@ -43,7 +43,7 @@ stdenv.mkDerivation rec {
     hostname
     python2 # update-compat.py
     python3 time # coq-makefile timing tools
-    dune
+    dune_2
   ]
   ++ (with ocamlPackages; [ ocaml findlib num ])
   ++ optionals buildIde [
diff --git a/dev/nixpkgs.nix b/dev/nixpkgs.nix
index 677377f868..b24a5dce47 100644
--- a/dev/nixpkgs.nix
+++ b/dev/nixpkgs.nix
@@ -1,4 +1,4 @@
 import (fetchTarball {
-  url = "https://github.com/NixOS/nixpkgs/archive/f4ad230f90ef312695adc26f256036203e9c70af.tar.gz";
-  sha256 = "0cdd275dz3q51sknn7s087js81zvaj5riz8f29id6j6chnyikzjq";
+  url = "https://github.com/NixOS/nixpkgs/archive/8d3199b549b9e5fa734d9f6c5f47a0c975bfce59.tar.gz";
+  sha256 = "0xvwm3awln83jq74vipmj2hqvhardni379ypn39950hza64y2180";
 })
```

However, when running `git clean -dfx && make -f Makefile.dune world`, I get multiple `ld: cannot find -lnums` errors.

`ocamlfind -query num` reports:
```
/nix/store/z0s681hn2pqh8lhxjyhc3z95sy30wj54-ocaml4.07.1-num-1.1/lib/ocaml/4.07.1/site-lib/num
```

@ejgallego @vbgl: Any idea what could be going on?

The problem doesn't occur when building with `make` within the same `nix-shell`.

**Comment by ejgallego** [comment]
Indeed this is a mess involving many components I do believe [has to do with some hacks used when num was distributed] I will locate the upstream bug reports.

What is your ocaml version, dune, and ocamlfind ? Does updating to latest packages help?

What are the contents of 
```
/nix/store/z0s681hn2pqh8lhxjyhc3z95sy30wj54-ocaml4.07.1-num-1.1/lib/ocaml/4.07.1/site-lib/num
```
? Is that directory empty [if so, that's the problem]

It would also help to see `_build/log` call

So indeed for example in Ubuntu + OPAM the num directory is empty but I think `ocamlfind` does some hack to make things work.

**Comment by ejgallego** [comment]
Some more information about what could be possibly wrong https://github.com/ocaml/opam-repository/issues/10925

But indeed I am leaning towards thinking that this could be a nix packaging issue as num really has a weird setup, what happens if you upgrade the nix pkgs ref but not dune?

For reference see num's standard META file:
```
requires = "num.core"
requires(toploop) = "num.core,num-top"
version = "1.3"
description = "Arbitrary-precision rational arithmetic"
package "core" (
  directory = "^"
  version = "1.3"
  browse_interfaces = ""
  archive(byte) = "nums.cma"
  archive(native) = "nums.cmxa"
  plugin(byte) = "nums.cma"
  plugin(native) = "nums.cmxs"
)
```
`directory` is `"^"` which denotes the ocaml root, likely this doesn't interact well with nums packaging.

**Comment by Zimmi48** [comment]
AFAICT this update does not include any change to num itself, nor to OCaml... The OCaml version used is 4.07.1 (the default now in nixpkgs), dune version is 2.1.0, and ocamlfind version is 1.8.1. The `site-lib/` directory of `num` is far from empty, and the `META` file seems more standard than the one you showed. Attached is the [`_build/log` file](https://github.com/coq/coq/files/3998706/log.txt).

I cannot simply test updating the nixpkgs ref without updating Dune because Coq requires Dune 2 to build now, but I did checkout the Coq commit before merging the PR updating the dependency on Dune, and tested doing the build with Dune 1.11.4 and the latest nixpkgs version and it works... I did test again this exact same commit of Coq with the same version of nixpkgs but Dune 2.1.0 and it gives the same error...
